### PR TITLE
feat(kafka): simplify creating kafka consumer and producer

### PR DESCRIPTION
**Pull request by RappC**
Still work in progress...

### CODE REVIEW

File: sda-commons-server-kafka/src/test/java/org/sdase/commons/server/kafka/KafkaBundleWithConfigIT.java [hunk 1]
```diff
@@ -880,6 +880,74 @@ public void shouldSetConsumerNameCorrectlyWithConsumerConfigFromYamlWithExplicit
     }
   }
 
+  @Test
+  public void test1() {
```
Comment (Fox32) [inline sda-commons-server-kafka/src/test/java/org/sdase/commons/server/kafka/KafkaBundleWithConfigIT.java:884]: Provide a better test name?

File: sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java [hunk 2]
```diff
@@ -383,19 +457,88 @@ private ExpectedTopicConfiguration createTopicDescription(TopicConfig c) {
...
+    ProducerRegistration<K, V> producerRegistration =
+        ProducerRegistration.<K, V>builder()
+            .forTopic(topicName)
+            .withDefaultProducer()
+            .withKeySerializer((Serializer<K>) getSerializer(keyClass, objectMapper))
+            .withValueSerializer((Serializer<V>) getSerializer(valueClass, objectMapper))
+            .build();
+
+    return registerProducer(producerRegistration);
+  }
```
Comment (Fox32) [inline sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java:486]: 👍  I like this

**Review by Fox32 — COMMENTED** [review]
@RappC Please run spotlessApply locally, the formatting is not correct

Any plans to continue with this PR?

File: sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java [hunk 3]
```diff
@@ -383,19 +457,88 @@ private ExpectedTopicConfiguration createTopicDescription(TopicConfig c) {
...
+   * Creates and registers a message producer with a default producer configuration
+   */
+  @SuppressWarnings("unchecked")
+  public <K, V> MessageProducer<K, V> createAndRegisterProducer(
+      Class<K> keyClass, Class<V> valueClass, ObjectMapper objectMapper, String topicName) {
+
+    ProducerRegistration<K, V> producerRegistration =
+        ProducerRegistration.<K, V>builder()
+            .forTopic(topicName)
+            .withDefaultProducer()
```
Comment (dhenneke) [inline sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java:480]: I would propose to not add a shortcut for the `withDefaultProducer` option. If the producer is not controlled from the config.yml, a user of the service has no chance to change settings if they are needed in the deployment.

File: sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java [hunk 4]
```diff
@@ -531,7 +674,25 @@ private void applyForcedConfigFromStrategy(
     consumerConfig.setConfig(newConfig);
   }
 
-  /** Initial checks. Configuration mus be initialized */
+  private static Serializer<?> getSerializer(Class<?> clazz, ObjectMapper objectMapper) {
```
Comment (dhenneke) [inline sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java:677]:
```suggestion
  private static Serializer<T> getSerializer(Class<T> clazz, ObjectMapper objectMapper) {
```

I didn't tested it, but I would assume that we could get rid of the unchecked warnings above.

File: sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java [hunk 5]
```diff
@@ -383,19 +457,88 @@ private ExpectedTopicConfiguration createTopicDescription(TopicConfig c) {
...
+            .build();
+
+    return registerProducer(producerRegistration);
+  }
+
+  /**
+   * Creates and registers a message producer with the given serializers and producer configuration
+   */
+  @SuppressWarnings("unchecked")
+  public <K, V> MessageProducer<K, V> createAndRegisterProducer(
```
Comment (dhenneke) [inline sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java:511]: The documentation is a bit short. Javadoc will complain that neither the properties nor the return type are documented.

**Review by dhenneke — COMMENTED** [review]
It would be nice if you could improve at the Readme.md of the Kafka bundle and document your new methods as recommended ones. Using the Kafka bundle is not trivial and a good documentation could help others to not get confused with so many options to choose from.

File: sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java [hunk 6]
```diff
@@ -383,19 +457,88 @@ private ExpectedTopicConfiguration createTopicDescription(TopicConfig c) {
...
+
+  /**
+   * Creates and registers a message producer with a default producer configuration
+   */
+  @SuppressWarnings("unchecked")
+  public <K, V> MessageProducer<K, V> createAndRegisterProducer(
+      Class<K> keyClass, Class<V> valueClass, ObjectMapper objectMapper, String topicName) {
+
+    ProducerRegistration<K, V> producerRegistration =
+        ProducerRegistration.<K, V>builder()
```
Comment (dhenneke) [inline sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/KafkaBundle.java:478]:
```suggestion
        ProducerRegistration.builder()
```

When I remember the changes we did for DW2 correctly, this shouldn't be necessary anymore.

**Pull request closed without merging**